domain/snyk/delta/code_matcher.go

```diff
@@ -0,0 +1,277 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package delta
+
+import (
+	"errors"
+	"github.com/adrg/strutil"
+	"github.com/adrg/strutil/metrics"
+	"math"
+	"path/filepath"
+	"strings"
+)
+
+var _ Matcher = (*CodeMatcher)(nil)
+
+type IssueConfidence struct {
+	BaseUUID           string
+	IssueIDResultIndex int
+	Confidence         float64
+}
+
+type DeduplicatedIssuesToIDs map[int]Identity
+type IssueConfidenceList []IssueConfidence
+type Identity struct {
+	IdentityID string
+	Confidence float64
+}
+
+var weights = struct {
+	MinimumAcceptableConfidence float64
+	FilePositionDistance        float64
+	RecentHistoryDistance       float64
+	FingerprintConfidence       float64
+	PathSimilarity              float64
+	LineSimilarity              float64
+
+	DirSimilarity           float64
+	FileNameSimilarity      float64
+	FileExtensionSimilarity float64
+}{
+	MinimumAcceptableConfidence: 0.4,
+	FilePositionDistance:        0.3,
+	RecentHistoryDistance:       0.2,
+	FingerprintConfidence:       0.5,
+
+	PathSimilarity: 0.8,
+	LineSimilarity: 0.2,
+
+	DirSimilarity:           0.5,
+	FileNameSimilarity:      0.3,
+	FileExtensionSimilarity: 0.2,
+}
+
+type CodeMatcher struct {
+}
+
+func (_ CodeMatcher) Match(baseIssueList, currentIssueList []Identifiable) ([]Identifiable, error) {
+	if len(currentIssueList) == 0 || len(baseIssueList) == 0 {
+		return nil, errors.New("base or current issue list is empty")
+	}
+
+	strongMatchingIssues := make(map[string]IssueConfidence)
+	existingAssignedIds := make(map[string]bool)
+
+	for index, issue := range currentIssueList {
+		if len(issue.GetGlobalIdentity()) > 0 {
+			existingAssignedIds[issue.GetGlobalIdentity()] = true
+			continue
+		}
+		findMatch(issue, index, baseIssueList, strongMatchingIssues)
+	}
+
+	finalResult := deduplicateIssues(strongMatchingIssues)
+
+	// Assign identities found to results
+	for i, identity := range finalResult {
+		if !existingAssignedIds[identity.IdentityID] {
+			currentIssueList[i].SetGlobalIdentity(identity.IdentityID)
+		}
+	}
+
+	// Assign UUIDs for any new issues
+	return currentIssueList, nil
+}
+
+func findMatch(issue Identifiable, index int, baseIssueList []Identifiable, strongMatchingIssues map[string]IssueConfidence) {
+	matches := findMatches(issue, index, baseIssueList)
+
+	for _, match := range matches {
+		if existing, exists := strongMatchingIssues[match.BaseUUID]; !exists || existing.Confidence < match.Confidence {
+			strongMatchingIssues[match.BaseUUID] = match
+		}
+	}
+}
+
+func findMatches(currentIssue Identifiable, index int, baseIssues []Identifiable) IssueConfidenceList {
+	similarIssues := make(IssueConfidenceList, 0)
+
+	for _, baseIssue := range baseIssues {
+		if baseIssue.RuleId() != currentIssue.RuleId() {
+			continue
+		}
+
+		filePositionDistance := filePositionConfidence(baseIssue, currentIssue)
```

Comment:
nitpick: why not `filePositionDistance` as function name (or a different variable name)?